pkg/policy/mapstate.go

```diff
@@ -273,6 +284,224 @@ func (msm *mapStateMap) ForEach(f func(Key, MapStateEntry) bool) bool {
 	return true
 }
 
+func (msm *mapStateMap) forKey(k Key, f func(Key, MapStateEntry) bool) bool {
+	e, ok := msm.entries[k]
+	if ok {
+		return f(k, e)
+	}
+	stacktrace := hclog.Stacktrace()
+	log.Errorf("Missing MapStateEntry for key: %v. Stacktrace: %s", k, stacktrace)
+	return true
+}
+
+// ForEachNarrowerCIDRAncestor iterates over narrower port/proto's and broader CIDRs in the trie.
+// Equal port/protos or identities are not included.
+func (msm *mapStateMap) ForEachNarrowerCIDRAncestor(key Key, prefixes []netip.Prefix, f func(Key, MapStateEntry) bool) {
+	msm.trie.Descendants(key.PrefixLength(), key, func(_ uint, lpmKey bitlpm.Key[policyTypes.Key], idSet IDSet) bool {
+		// k is the key from trie with 0'ed ID
+		k := lpmKey.Value()
+
+		// Descendants iterates over the key itself if it is in the trie, skip it
```

Comment:
It can iterate over a key of another equal port-protocol. In which case they should both be inserted, but no shadow-key is necessary.

Reply:
Comment was a bit wrong, it now reads:
```
		// Descendants iterates over equal port/proto, caller expects to see only narrower keys so skip it
```